lib/modules/dosomething/dosomething_global/dosomething_global.module

```diff
@@ -14,77 +14,60 @@ define('DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE', 'en-global');
-    $args = explode('/', $path);
-
-    // Only apply logic to Global homepage
-    if (drupal_is_front_page()) {
-      if (!empty($path) && $args[0] != 'node') {
-        return;
-      }
-    }
-
-    $country_code = dosomething_settings_get_geo_country_code();
-    $language = dosomething_global_convert_country_to_language($country_code);
-    // No matching language, serve the global page
-    if ($language == NULL) {
-      return;
-    }
+  $languages = language_list();
+  if (empty($languages[DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE])) {
+    watchdog('dosomething_global', "Can't load %lang language",
+      array('%lang' => $global_lang_code), WATCHDOG_ERROR);
+    return;
+  }
 
-    // Get users local path prefix and verify we support it
-    $path_prefix = dosomething_global_get_prefix_for_language($language);
-    if ($path_prefix == NULL) {
-      return;
-    }
+  global $user;
+  $url_variables = dosomething_global_get_url_variables();
+  $user_variables = dosomething_global_get_user_variables();
+  $node_variables = dosomething_global_get_node_variables();
 
-    // Redirect user
-    drupal_goto($path_prefix . '/' . current_path());
+  // Verify we're dealing with a node edit with no translation specification in the URL
+  if ($node_variables['is_node'] && $node_variables['is_node_edit'] && $node_variables['node_edit_language'] == NULL) {
+    dosomething_global_redirect_node_edit($node_variables['node'], $user_variables['user_language']);
+    return;
+  }
+  // If we already have a path prefix,
+  // or the user is a global english speaker,
+  // all logic after this statement should not run.
+  if ($url_variables['url_path_prefix'] != dosomething_global_get_prefix_for_language(DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE)) {
+    return;
+  }
+  else if($user_variables['user_language'] == DOSOMETHING_GLOBAL_DEFAULT_LANG_CODE) {
+    return;
   }
 
-  // Verify we're dealing with a node edit with no translation specification in the URL
-  if (arg(0) == "node" && is_numeric(arg(1)) && arg(2) == "edit" && null == arg(3)) {
-    // Load the page node and user
-    $nid = arg(1);
-    $node = node_load($nid);
-
-    // Load the languages being used
-    $src_language = $node->language;
-    $user_language = $user->language ?: 'en';
-
-    // If we're not working on a translation, no need to re-direct.
-    if ($src_language != $user_language) {
-      // Determines if the url needs the /add/source/ component
-      $new_translation_prefix = isset($node->translations->data[$user_language]) ? '' : 'add/' . $src_language . '/';
-
-      // Append the user language
-      $lang_append = $new_translation_prefix . $user_language;
-
-      // Build complete URL and redirect
-      $new_url = str_replace(' ', '', '/node/' . $nid . '/edit/' . $lang_append);
-      drupal_goto($new_url);
-    }
+  // Configure the correct path prefix
+  $url_variables['url_path_prefix'] = dosomething_global_get_prefix_for_language($user_variables['user_language']);
```

Comment:
can this prefix stuff be refactored a bit to use the function @sbsmith86 just wrote? `dosomething_global_get_current_prefix()`